OCPBUGS-99480: Harden Gateway API docs against route hijacking (#117235)
openshift/openshift-docs: modules/gateway-listener-configuration-reference.adoc

Defines the network port and protocol. For example, `protocol: HTTP` accepts HTTP traffic on port `80`, and `protocol: HTTPS` accepts HTTPS traffic on port `443`.

`listeners.hostname`::

> Defines the hostnames that the listener matches for incoming requests. For example, it can match only requests for hostnames ending in `<example_domain.tld>` (such as `www.<example_domain.tld>` or `*.gwapi.<example_domain.tld>`). Always set a hostname on each listener. If no hostname is specified, any route that can attach to the listener can claim arbitrary hostnames, which can allow hostname or domain hijacking.

Reviewer: unless the user has a good reason to not set a hostname, like if they want to define a catch-all route.

Anyway, on managed cloud environments (AWS, GCP, Azure), not setting a hostname on the listener will cause a DNSRecord to not be created (not a problem/bug, but the Gateway will not have an automatic DNS entry created)

`listeners.tls`::
Specifies the TLS settings for secure communication, including the mode (currently, only `Terminate` is supported on {product-title}) and the Kubernetes secret containing the certificate keypair. For example, TLS is terminated at the gateway using a certificate stored in a Kubernetes `Secret` called `<listener_cert>`. You must ensure that the specified Kubernetes secret exists before you create the `Gateway` CR.

`listeners.allowedRoutes`::

> Controls which route resources can attach to this listener. For example, an HTTP listener might only allow `HTTPRoute` resources from namespaces that have the `env: "dev"` label (using `Selector`). Setting `allowedRoutes.namespaces.from: Same` is not supported; routes from the same namespace as the gateway are always allowed. Do not use `.spec.listeners[].allowedRoutes.namespaces.from=All` because this is insecure and can allow hostname or domain hijacking.

Reviewer 2: ```suggestion
Controls which route resources can attach to this listener. For example, an HTTP listener might only allow `HTTPRoute` resources from namespaces that have the `env: "dev"` label (using `Selector`). Setting `allowedRoutes.namespaces.from: Same` is not supported; the default behaviour without the use of allowedRoutes allows all routes from the same namespace as the gateway to attach. Do not use `.spec.listeners[].allowedRoutes.namespaces.from=All` because this is insecure and can allow hostname or domain hijacking.
```
maybe this can be reworded but I want to mention that this is the default behaviour hence `allowedRoutes.namespaces.from: Same` does not work

Author: @rhamini3, what does it mean to "not work"?

If a user sets that, the gateway doesn't work at all?
If a user sets that, the behavior is the same as if it was not set?
If a user sets that, the behavior is undefined and we don't know what happens?

Reviewer 2: > If a user sets that, the behavior is the same as if it was not set?
it is this
